spec.md

```diff
@@ -1598,6 +1605,58 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 | Invalid `starting_token` | 10 ABORTED | Indicates that `starting_token` is not valid. | Caller SHOULD start the `ListSnapshots` operation again with an empty `starting_token`. |
 
 
+#### `ControllerExpandVolume`
+
+A Controller plugin MUST implement this RPC call if plugin has `EXPAND_VOLUME` capability. 
+This RPC allows the CO to expand size of underlying volume.
+
+This call can be made by the CO during any time in the lifecycle of the volume after creation.
+If plugin has `EXPAND_VOLUME` node capability, then `NodeExpandVolume` MUST be called after `ControllerExpandVolume` has completed.
+If volume can not be expanded when in-use - plugin should return `FAILED_PRECONDITION` error and CO should retry only after:
```

Comment:
This means that the CO cannot know if the plugin is capable of expanding the specified volume beforehand, and thus the CO needs to try the call first and relies on the response to determine if it can expand a volume that is in use.

Instead of making the CO do some try-and-error, can we introduce e.g., a volume capability to specify that if a volume is able to be expanded, and if it is allowed to be expanded when in use, etc?

Reply:
By capability I think you mean - if volume can be expanded while in-use or not, right? This PR already adds a controller `EXPAND_VOLUME` capability which CO can use.

In practice - CO does some try-and-error has worked well enough. Take `DeleteVolume` RPC call for example. The error condition here is almost same as `Volume In Use` error condition there. I think - we should let specs evolve rather than try and encode every possible combination and capability(only to get them wrong).

Reply:
No I'm talking about volume capability, not controller/node capability or plugin capability.

For example, a plugin might want to support more than one fs type, but not all fs types allow expanding volumes while being mounted. So it seems to me that this is more appropriate to be specified on a per-volume basis rather than a per-plugin basis.

Re: Try-and-error. IMO the CO should try to avoid errors as hard as possible, so the error handling is just a safe guard when something unexpected happens. It seems to me that a try-and-error approach would make the CO hard to maintain its view to the volume states, and also potentially reduce user experience. Otherwise, we could go as bold as just removing all capabilities and let the CO to try calling all calls and read the gRPC errors, right? ;)

Reply:
Sorry I misunderstood your comment about capabilities. Yes - that is something we can definitely add to the API. I need to think more.

From our k8s experience, most volume types support online/in-use expansion and but there are definitely exceptions to the norm. There are two different definitions of in-use though. Some volume types are forbidden to be resized when attached to a node, regardless of file system on it. That is a control-plane restriction.

In some other cases, it may be a file system restriction. Is there any modern file system you had in mind that requires volume to be unmounted for expansion?

Reply:
I also like the idea of using `VolumeCapability` at per volume basis. This also allows the CO to call `ValidateVolumeCapability` to decide if it can expand volume based on the state of the volume. For instance, we can introduce two types of volume expansion capabilities. One type allows in-use volume expansion, the other does not.